### PR TITLE
2850 refactor to remove api23 kludges and switches

### PR DESCRIPTION
**Pull request by seabornlee**
- Fix test
- Fix test
- Ignore failure test
- Remove version check code
- Use Java 8 stream API
- Use original web3j lib

### CODE REVIEW

**Comment by JamesSmartCell** [comment]
@seabornlee some of these updates probably might not help readability of the source; where there's a for loop iterating through eg a list or map maybe let's keep it that way for readability/maintainability, but if its a collection then use the streams.

What do you think?

**Comment by seabornlee** [comment]
Can you comment on several points which you prefer to keep?

**Comment by JamesSmartCell** [comment]
> Can you comment on several points which you prefer to keep?

Sure, basically anywhere where a 'for' loop would make the code cleaner, readable and maintainable. It would be easier to highlight the areas where this does improve the code rather than the other way round because I think about 90% of the time a 'for' loop is better. I find it very difficult to read or work on the code where this stream style is used, unless it used to traverse a structure that's already a collection, where it appears natural.

File: app/src/main/java/com/alphawallet/app/entity/CovalentTransaction.java [hunk 1]
```diff
@@ -95,34 +97,26 @@ public Map<String, Param> getParams() throws Exception
     public static EtherscanTransaction[] toEtherscanTransactions(CovalentTransaction[] transactions, NetworkInfo info)
     {
         List<EtherscanTransaction> converted = new ArrayList<>();
-        for (CovalentTransaction tx : transactions)
-        {
+        Arrays.stream(transactions).forEach(tx -> {
             try
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/CovalentTransaction.java:101]: Please revert to 'for' loop as it's simpler and cleaner

File: app/src/main/java/com/alphawallet/app/entity/CovalentTransaction.java [hunk 2]
```diff
@@ -95,34 +97,26 @@ public Map<String, Param> getParams() throws Exception
...
     {
         if (log_events == null || log_events.length == 0) return "";
-        for (LogEvent le : log_events)
-        {
-            if (le.sender_address != null)
-            {
-                return le.sender_address;
-            }
-        }
+        return Arrays.stream(log_events).filter(le -> le.sender_address != null).findFirst().map(le -> le.sender_address).orElse("");
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/CovalentTransaction.java:118]: Please revert to 'for' loop here.

File: app/src/main/java/com/alphawallet/app/entity/EventSync.java [hunk 3]
```diff
@@ -397,10 +398,7 @@ public Pair<Integer, Pair<HashSet<BigInteger>, HashSet<BigInteger>>> processTran
         HashSet<BigInteger> sendTokenIds = token.processLogsAndStoreTransferEvents(sentLogs, transferEvent, txHashes, realm);
 
         //register Transaction fetches
-        for (String txHash : txHashes)
-        {
-            TransactionsService.addTransactionHashFetch(txHash, token.tokenInfo.chainId, token.getWallet());
-        }
+        txHashes.forEach(txHash -> TransactionsService.addTransactionHashFetch(txHash, token.tokenInfo.chainId, token.getWallet()));
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/EventSync.java:401]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/GasPriceSpread.java [hunk 4]
```diff
@@ -42,14 +45,7 @@ public int getEntrySize()
...
     {
-        for (TXSpeed txs : TXSpeed.values())
-        {
-            GasSpeed2 gs = fees.get(txs);
-            if (gs != null) return gs;
-        }
-
-        //Should not reach this
-        return null;
+        return Arrays.stream(TXSpeed.values()).map(fees::get).filter(Objects::nonNull).findFirst().orElse(null);
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/GasPriceSpread.java:48]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/GasPriceSpread.java [hunk 5]
```diff
@@ -246,16 +242,15 @@ private void setComponents(Context ctx, Map<Integer, EIP1559FeeOracleResult> res
         fees.put(TXSpeed.SLOW, new GasSpeed2(ctx.getString(R.string.speed_slow), SLOW_SECONDS, new EIP1559FeeOracleResult(result.get(result.size()-1))));
 
         //now de-duplicate
-        for (TXSpeed txs : TXSpeed.values())
-        {
+        Arrays.stream(TXSpeed.values()).forEach(txs -> {
             GasSpeed2 gs = fees.get(txs);
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/GasPriceSpread.java:246]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/GasTransactionResponse.java [hunk 6]
```diff
@@ -114,10 +115,7 @@ public void truncatePriceRange()
     {
         arrayResult = new SparseArray<>();
         SparseArray<Float> priceSet = new SparseArray<>();
-        for (String price : result.keySet())
-        {
-            priceSet.put(Integer.valueOf(price), result.get(price));
-        }
+        result.keySet().forEach(price -> priceSet.put(Integer.parseInt(price), result.get(price)));
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/GasTransactionResponse.java:118]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/nftassets/NFTAsset.java [hunk 7]
```diff
@@ -177,15 +168,7 @@ public String getAnimation()
...
-        for (String key : IMAGE_DESIGNATORS)
-        {
-            if (assetMap.containsKey(key))
-            {
-                return Utils.parseIPFS(assetMap.get(key));
-            }
-        }
-
-        return "";
+        return Arrays.stream(IMAGE_DESIGNATORS).filter(assetMap::containsKey).findFirst().map(key -> Utils.parseIPFS(assetMap.get(key))).orElse("");
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/nftassets/NFTAsset.java:171]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/nftassets/NFTAsset.java [hunk 8]
```diff
@@ -301,15 +277,8 @@ private boolean validJSONString(String value)
...
     private String getSVGOverride()
     {
-        for (String key : SVG_OVERRIDE)
-        {
-            if (assetMap.containsKey(key) && assetMap.get(key).toLowerCase().endsWith("svg"))
-            {
-                return Utils.parseIPFS(assetMap.get(key));
-            }
-        }
+        return Arrays.stream(SVG_OVERRIDE).filter(key -> assetMap.containsKey(key) && assetMap.get(key).toLowerCase().endsWith("svg")).findFirst().map(key -> Utils.parseIPFS(assetMap.get(key))).orElse("");
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/nftassets/NFTAsset.java:280]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Ticket.java [hunk 9]
```diff
@@ -61,12 +63,7 @@ public String getFullBalance() {
 
     @Override
     public Map<BigInteger, NFTAsset> getTokenAssets() {
-        Map<BigInteger, NFTAsset> assets = new HashMap<>();
-        for (BigInteger tokenId : balanceArray)
-        {
-            assets.put(tokenId, new NFTAsset(tokenId));
-        }
-        return assets;
+        return balanceArray.stream().collect(Collectors.toMap(tokenId -> tokenId, NFTAsset::new, (a, b) -> b));
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Ticket.java:66]: Please revert to 'for' loop. This is a bit impenetrable what's going on in here, would never be sure exactly what this does without expanding the code back out.

File: app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Ticket.java [hunk 10]
```diff
@@ -171,7 +165,7 @@ public boolean hasArrayBalance()
     public List<BigInteger> getNonZeroArrayBalance()
     {
         List<BigInteger> nonZeroValues = new ArrayList<>();
-        for (BigInteger value : balanceArray) if (value.compareTo(BigInteger.ZERO) != 0 && !nonZeroValues.contains(value)) nonZeroValues.add(value);
+        balanceArray.stream().filter(value -> value.compareTo(BigInteger.ZERO) != 0 && !nonZeroValues.contains(value)).collect(Collectors.toList());
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Ticket.java:168]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Token.java [hunk 11]
```diff
@@ -253,11 +254,7 @@ public boolean checkRealmBalanceChange(RealmToken realmToken)
         if (lastTxTime > realmToken.getLastTxTime()) return true;
         if (!currentState.equals(balance.toString())) return true;
         //check balances
-        for (NFTAsset a : tokenBalanceAssets.values())
-        {
-            if (!a.needsLoading() && !a.requiresReplacement()) return true;
-        }
-        return false;
+        return tokenBalanceAssets.values().stream().anyMatch(a -> !a.needsLoading() && !a.requiresReplacement());
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Token.java:257]: Please revert to 'for' loop.

File: app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Token.java [hunk 12]
```diff
@@ -475,15 +472,12 @@ private void removeRealmBalance(Realm realm, HashSet<BigInteger> removedTokens)
...
+        HashSet<BigInteger> storedBalance;
         RealmResults<RealmNFTAsset> results = realm.where(RealmNFTAsset.class)
                 .like("tokenIdAddr", databaseKey(this) + "-*", Case.INSENSITIVE)
                 .findAll();
 
-        for (RealmNFTAsset t : results)
-        {
-            storedBalance.add(new BigInteger(t.getTokenId()));
-        }
+        storedBalance = results.stream().map(t -> new BigInteger(t.getTokenId())).collect(Collectors.toCollection(HashSet::new));
```
Comment (JamesSmartCell) [inline app/src/main/java/com/alphawallet/app/entity/tokens/ERC721Token.java:480]: Please revert to 'for' loop.

**Review by JamesSmartCell — CHANGES_REQUESTED** [review]
I got a little way into this, then realised that pretty much anywhere where we can do:

```for(data : dataset)``` the code is clearer and more maintainable if we leave it as a for loop.

The streams are good in many places, esp where the structure is like this:

```for (int i=0; i<x; i++)```

Definitely improves that code.

However, please revert to ```for``` anywhere where we use the for loop for traversing a dataset. I think with streams and the short notation we're going to run into maintenance problems further down the line.

**Pull request closed without merging**